tests/auth_tests/test_migrations.py

```diff
@@ -152,3 +154,36 @@ def test_user_keeps_same_permissions_after_migrating_backward(self):
         user = User._default_manager.get(pk=user.pk)
         for permission in [self.default_permission, self.custom_permission]:
             self.assertTrue(user.has_perm('auth_tests.' + permission.codename))
+
+    def test_migrate_with_existing_permission_from_previously_concrete_model(self):
+        """
+        Permissions may already exist:
+
+        - Old workaround was to manually create permissions for proxy models.
+        - Model may have been concrete and then converted to proxy. (#30351)
+
+        Do not crash in this case. Output a reminder to audit relevant
+        permissions.
+        """
+        proxy_model_content_type = ContentType.objects.get_for_model(
+            Proxy, for_concrete_model=False)
+
+        Permission.objects.create(
+            content_type=proxy_model_content_type,
+            codename='add_proxy',
+            name='Can add proxy',
+        )
+        Permission.objects.create(
+            content_type=proxy_model_content_type,
+            codename='display_proxys',
+            name='May display proxys information',
+        )
+
+        stream = io.StringIO()
+        old_stdout = sys.stdout
+        sys.stdout = stream
+
+        update_proxy_permissions.update_proxy_model_permissions(apps, None)
+
+        sys.stdout = old_stdout
+        self.assertIn('A problem arose migrating proxy model permissions', stream.getvalue())
```

Comment:
We can simplify this with `captured_stdout()`, e.g.
```diff
diff --git a/tests/auth_tests/test_migrations.py b/tests/auth_tests/test_migrations.py
index 4b4fe4bda2..56494d58d1 100644
--- a/tests/auth_tests/test_migrations.py
+++ b/tests/auth_tests/test_migrations.py
@@ -6,6 +6,7 @@ from django.apps import apps
 from django.contrib.auth.models import Permission, User
 from django.contrib.contenttypes.models import ContentType
 from django.test import TestCase
+from django.test.utils import captured_stdout
 
 from .models import Proxy, UserProxy
 
@@ -178,12 +179,6 @@ class ProxyModelWithSameAppLabelTests(TestCase):
             codename='display_proxys',
             name='May display proxys information',
         )
-
-        stream = io.StringIO()
-        old_stdout = sys.stdout
-        sys.stdout = stream
-
-        update_proxy_permissions.update_proxy_model_permissions(apps, None)
-
-        sys.stdout = old_stdout
-        self.assertIn('A problem arose migrating proxy model permissions', stream.getvalue())
+        with captured_stdout() as stdout:
+            update_proxy_permissions.update_proxy_model_permissions(apps, None)
+            self.assertIn('A problem arose migrating proxy model permissions', stdout.getvalue())
```

Reply:
Good. It's almost as if someone needed to test this sort of thing before. 😀